python/dgl/graphbolt/item_sampler.py

```diff
@@ -167,19 +176,102 @@ def _calculate_offsets(self, buffer):
         return torch.tensor(offsets)
 
     def __iter__(self):
+        worker_info = torch.utils.data.get_worker_info()
+        if worker_info is not None:
+            num_workers = worker_info.num_workers
+            worker_id = worker_info.id
+        else:
+            num_workers = 1
+            worker_id = 0
         buffer = None
-        num_items = len(self._item_set)
+        if not self._distributed:
+            num_items = len(self._item_set)
+            start_offset = 0
+        else:
+            total_count = len(self._item_set)
+            big_batch_size = self._num_replicas * self._batch_size
+            big_batch_count = total_count // big_batch_size
+            big_batch_remain = total_count % big_batch_size
+            batch_remain = big_batch_remain % self._batch_size
+            last_batch_count = big_batch_remain // self._batch_size
+            if self._rank < last_batch_count:
+                last_batch = self._batch_size
+            elif self._rank == last_batch_count:
+                last_batch = batch_remain
+            else:
+                last_batch = 0
+            num_items = big_batch_count * self._batch_size + last_batch
+            start_offset = (
+                big_batch_count * self._batch_size * self._rank
+                + min(self._rank * self._batch_size, big_batch_remain)
+            )
+            if not self._drop_uneven_inputs or (
+                not self._drop_last and last_batch_count == self._num_replicas
+            ):
+                num_evened_items = num_items
+                if num_workers > 1:
```

Comment:
Roughly looking there are a lot of repeat code here, try to merge them and make the code cleaner.